File: project/Examples/Examples/PP2E/Integrate/Embed/Inventory/WithDbase/order-func.c
/* run embedded module-function validations */
/* reuse python file-load routine: returns list of tuples, */
/* so use Python PyObject object access functions to process */
/* also see abstract.h and lower-level Python object .h files */

#include <ppembed.h>
#include <stdio.h>
#include <string.h>

run_user_validation(char *ofile)
{
    int i, status;                /* should also check status everywhere */
    char *errors, *warnings;
    PyObject *results, *orders, *nextorder=NULL;
    Py_Initialize();
    PyRun_SimpleString("import sys; sys.path.append('..')");
    PP_Run_Function("inventory", "load_orders", "O", &orders, "(s)", ofile);

    PP_Run_Function("inventory", "print_files", "", NULL, "()");    
    for (i=0; i < PyObject_Length(orders); i++) {                   /* len(x) */
        Py_XDECREF(nextorder);
        nextorder = PyObject_GetItem(orders, PyInt_FromLong(i));    /* x[i] */
        printf("\n%d ", i);
        PyObject_Print(nextorder, stdout, 0);
        printf("\n");

        status = PP_Run_Function(                /* validate2.validate(p,q,b) */
                         "validate2", "validate",  
                         "O",          &results,
                         "O",          nextorder);
        if (status == -1) {
            printf("Python error during validation.\n");
            PyErr_Print();  /* show traceback */
            continue;
        }
        PyArg_Parse(results, "(ss)", &warnings, &errors);
        printf("errors:   %s\n", strlen(errors)? errors : "none"); 
        printf("warnings: %s\n", strlen(warnings)? warnings : "none"); 
        Py_DECREF(results);  /* ok to free strings */
        PP_Run_Function("inventory", "print_files", "", NULL, "()");    
    }
}

main(int argc, char **argv)
{
    char ofile[64];
    if (argc == 1)
        strcpy(ofile, "Data/ordersfile.data");
    else {
        strcpy(ofile, "Data/");
        strcat(ofile, argv[1]);
    }
    run_user_validation(ofile);
}

File: project/Examples/Examples/PP2E/Integrate/Embed/Inventory/WithDbase/order-func.py
#!/usr/local/bin/python
# to test the embedded-code script, without the enclosing C app 

import sys   
sys.path.append('..')              # get validate2 from above

import validate2                   # get validation function
import inventory                   # inventory, buyers
import traceback                   # stack dump on errors

if len(sys.argv) == 1:
    ofile = 'Data/ordersfile.data'
else:
    ofile = 'Data/' + sys.argv[1]
orders = inventory.load_orders(ofile)
print orders

inventory.print_files()
for order in orders: 
    print '\n', orders.index(order), order
    try:
        (warnings, errors) = apply(validate2.validate, order)  
    except:
        print 'error in embedded function:'
        traceback.print_exc()
        continue
    print 'errors:  ', errors   or 'none'
    print 'warnings:', warnings or 'none'
    inventory.print_files()
